// @mosteast/common_eid/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var EID_common;
(function (EID_common) {
    // !!!
    // EID SHOULD NEVER BE CHANGED
    //
    // Once an EID is set and published we must never change it.
    // For instance `base_error` should always be 'e00' and
    // invalid_argument should always be 'e01'.
    // If new EID is needed just increment the code number.
    // !!!
    // Common EID starts with string 'e0',
    // custom app EID should start with string 'e'.
    EID_common["base_error"] = "e00";
    EID_common["invalid_argument"] = "e01";
    EID_common["lack_config"] = "e02";
    EID_common["lack_path"] = "e03";
    EID_common["lack_argument"] = "e04";
    EID_common["permission_denied"] = "e05";
    EID_common["bad_connection"] = "e06";
    EID_common["invalid_state"] = "e07";
    EID_common["invalid_credential"] = "e08";
    EID_common["too_frequent"] = "e09";
    EID_common["invalid_configuration"] = "e010";
    EID_common["invalid_environment_variable"] = "e011";
})(EID_common = exports.EID_common || (exports.EID_common = {}));
